router/fastdp.go

```diff
@@ -0,0 +1,1109 @@
+	if _, present := fastdp.missHandlers[vport.ID]; !present {
+		fastdp.makeBridgeVport(vport)
+	}
+
+	return nil
+}
+
+func (fastdp *FastDatapath) VportDeleted(ifindex int32, vport odp.Vport) error {
+	fastdp.lock.Lock()
+	defer fastdp.lock.Unlock()
+
+	// there might be flow rules that still refer to the id of
+	// this vport.  But we just allow them to expire.  Unless we
+	// want prompt migration of MAC addresses, that should be
+	// fine.
+	delete(fastdp.missHandlers, vport.ID)
+	fastdp.deleteSendToPort(bridgePortID{vport: vport.ID})
+	return nil
+}
+
+func (fastdp *FastDatapath) makeBridgeVport(vport odp.Vport) {
+	// Set up a bridge port for netdev and internal vports.  vxlan
+	// vports are handled separately, as they do not correspond to
+	// bridge ports (we set up the miss handler for them in
+	// getVxlanVportID).
+	typ := vport.Spec.TypeName()
+	if typ != "netdev" && typ != "internal" {
+		return
+	}
+
+	vportID := vport.ID
+
+	// Sending to the bridge port outputs on the vport:
+	fastdp.addSendToPort(bridgePortID{vport: vportID},
+		func(_ PacketKey, _ *fastDatapathLock) FlowOp {
+			return fastdp.odpActions(odp.NewOutputAction(vportID))
+		})
+
+	// Delete flows, in order to recalculate flows for broadcasts
+	// on the bridge.
+	checkWarn(fastdp.deleteFlows())
+
+	// Packets coming from the netdev are processed by the bridge
+	fastdp.missHandlers[vportID] = func(flowKeys odp.FlowKeys, lock *fastDatapathLock) FlowOp {
+		return fastdp.bridge(bridgePortID{vport: vportID},
+			flowKeysToPacketKey(flowKeys), lock)
+	}
+}
+
+func flowKeysToPacketKey(fks odp.FlowKeys) PacketKey {
+	eth := fks[odp.OVS_KEY_ATTR_ETHERNET].(odp.EthernetFlowKey).Key()
+	return PacketKey{SrcMAC: eth.EthSrc, DstMAC: eth.EthDst}
+}
+
+// The sendToPort map is read-only, so this method does the copy in
+// order to add an entry.
+func (fastdp *FastDatapath) addSendToPort(portID bridgePortID,
+	sender bridgeSender) {
+	sendToPort := map[bridgePortID]bridgeSender{portID: sender}
+	for id, sender := range fastdp.sendToPort {
+		sendToPort[id] = sender
+	}
+	fastdp.sendToPort = sendToPort
+}
+
+func (fastdp *FastDatapath) deleteSendToPort(portID bridgePortID) {
+	sendToPort := make(map[bridgePortID]bridgeSender)
+	for id, sender := range fastdp.sendToPort {
+		if id != portID {
+			sendToPort[id] = sender
+		}
+	}
+	fastdp.sendToPort = sendToPort
+}
+
+// Send a packet, creating a corresponding ODP flow rule if possible
+func (fastdp *FastDatapath) send(fops FlowOp, frame []byte,
+	lock *fastDatapathLock) {
+	// Gather the actions from actionFlowOps, execute any others
+	var dec *EthernetDecoder
+	flow := odp.NewFlowSpec()
+	createFlow := true
+
+	for _, xfop := range FlattenFlowOp(fops) {
+		switch fop := xfop.(type) {
+		case interface {
+			updateFlowSpec(*odp.FlowSpec)
+		}:
+			fop.updateFlowSpec(&flow)
+		case vetoFlowCreationFlowOp:
+			createFlow = false
+		default:
+			// A foreign flow op, so send the packet the
+			// normal way, decoding the packet lazily.
```

Comment:
> Why would you configure things this way?

Well, at a certain point in fastdp development, I had the code to manage the vxlan tunnel flows, but I hadn't written the full bridge implementation.  But I had something that worked, because I could run with the nascent fastdp Overlay and the pcap-based Bridge (on top of a normal linux bridge netdev).

Even now, it only takes a simple modification to the weave script to run with the fastdp Overlay and pcap-based Bridge (although with the bridge actually being provided by fastdp too).  And it works, demonstrating that the design is general enough to support this with no substantive code changes.  A proper understanding of the fastdp code should not exclude such possibilities.